File: pkg/dev_compiler/test/codegen/methods.dart
// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library methods;

class A {
  int x() => 42;

  int y(int a) {
    return a;
  }

  int z([num b]) => b;

  int zz([int b = 0]) => b;

  int w(int a, {num b}) {
    return a + b;
  }

  int ww(int a, {int b: 0}) {
    return a + b;
  }

  clashWithObjectProperty({constructor}) => constructor;
  clashWithJsReservedName({function}) => function;

  int get a => x();

  void set b(int b) {}

  int _c = 3;

  int get c => _c;

  void set c(int c) {
    _c = c;
  }
}

class Bar {
  call(x) => print('hello from $x');
}
class Foo {
  final Bar bar = new Bar();
}

test() {
  // looks like a method but is actually f.bar.call(...)
  var f = new Foo();
  f.bar("Bar's call method!");

  // Tear-off
  A a = new A();
  var g = a.x;

  // Dynamic Tear-off
  dynamic aa = new A();
  var h = aa.x;

  // Tear-off of object methods
  var ts = a.toString;
  var nsm = a.noSuchMethod;

  // Tear-off extension methods
  var c = "".padLeft;
  var r = (3.0).floor;
}

File: pkg/dev_compiler/test/codegen/expect/methods/methods.js
var methods;
(function (methods) {
  'use strict';
  class A extends dart.Object {
    A() {
      this._c = 3;
    }
    x() { return 42; }
    y(a) {
      return a;
    }
    z(b) {
      if (b === undefined) b = null;
      return dart.notNull(b)
    }
    zz(b) {
      if (b === undefined) b = 0;
      return b
    }
    w(a, opt$) {
      let b = opt$.b === undefined ? null : opt$.b;
      return dart.notNull(a + dart.notNull(b));
    }
    ww(a, opt$) {
      let b = opt$.b === undefined ? 0 : opt$.b;
      return a + b;
    }
    get a() { return this.x(); }
    set b(b) {
    }
    get c() { return this._c; }
    set c(c) {
      this._c = c;
    }
  }

  class Bar extends dart.Object {
    call(x) { return core.print(`hello from ${x}`); }
  }

  class Foo extends dart.Object {
    Foo() {
      this.bar = new Bar();
    }
  }

  // Function test: () → dynamic
  function test() {
    let f = new Foo();
    dart.dinvoke(f, "bar", "Bar's call method!");
  }

  // Exports:
  methods.A = A;
  methods.Bar = Bar;
  methods.Foo = Foo;
  methods.test = test;
})(methods || (methods = {}));
